File: solutions/elixir/pascals-triangle/1/lib/pascals_triangle.ex
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(num) do
    rows(num, [[1]])
  end

  defp rows(1, acc), do: Enum.reverse(acc)
  defp rows(num, [h | t]), do: rows(num - 1, [row(h, [1]), h | t])

  defp row([h], acc), do: [h | acc]

  defp row([h0, h1 | t], acc) do
    row([h1 | t], [h0 + h1 | acc])
  end
end

File: solutions/javascript/pascals-triangle/1/pascals-triangle.js
export const rows = (numRows) => {
  if (numRows === 0) return [];

  const triangle = [];
  triangle.push(getRow([]));
  for (let i = 1; i < numRows; i++) {
      triangle.push(getRow(triangle[i - 1]));
  }
  return triangle;
};

const getRow = (prevRow) => {
  if (prevRow.length === 0) return [1];

  const prevWithZero = [0, ...prevRow, 0];
  const row = [];
  for (let i = 0; i < prevWithZero.length - 1; i++) {
    row.push(prevWithZero[i] + prevWithZero[i + 1]);
  }
  return row;
}
